packages/bookshelf-include-count/lib/bookshelf-include-count.js

```diff
@@ -157,8 +226,71 @@ module.exports = function (Bookshelf) {
             });
 
             return attrs;
-        }
+        },
+
+        /**
+         * Instead of adding the counts in .fetch and .fetchAll,
+         * we need to do it in sync because Bookshelf doesn't call fetch for eagerRelations
+         * E.g. when trying to load counts on replies.count.likes, we wouldn't get an opportunity to load the counts on the replies relation.
+         */
+        sync: function (options) {
+            if (!options.method || (options.method !== 'insert' && options.method !== 'update')) {
+                this.addCounts.apply(this, arguments);
+            }
+
+            if (_debug.enabled('ghost-query')) {
+                debug('QUERY', this.query().toQuery());
+            }
+
+            // Call parent fetchAll
+            return modelProto.sync.apply(this, arguments);
+        },
+
+        // When save a modal, make sure we keep the already fetched counts alive because they will get
+        // removed before saving in the saving event (as they are not 'permitted')
+        // We need to use bluebird to override save (or things will break)
+        save: Promise.method(function save() {
+            // the count__ variables are not 'permitted' and will get removed after a save
+            // so this will make sure they are kept alive after a save (unless they are also still available after the save)
+
+            const savedAttributes = {};
+            const countRegex = /^(count)(__)(.*)$/;
+
+            for (const key of Object.keys(this.attributes)) {
+                const match = key.match(countRegex);
+                if (match) {
+                    savedAttributes[key] = this.attributes[key];
+                }
+            }
+            
+            return modelProto.save.apply(this, arguments).then((t) => {
+                // Set savedAttributes, but keep count__ variables if they stayed inside this.attributes
+                if (savedAttributes) {
+                    Object.assign(this.attributes, savedAttributes, this.attributes);
+                }
+                return t;
+            });
+        })
     });
 
     Bookshelf.Model = Model;
+
+    const collectionProto = Bookshelf.Collection.prototype;
+
+    const Collection = Bookshelf.Collection.extend({
+        addCounts,
+        sync: function () {
+            // For now, only apply this for eager loaded collections
+            this.addCounts.apply(this, arguments);
+
+            if (_debug.enabled('ghost-query')) {
```

Comment:
I get that you're following the existing convention, but I have no idea why we're using this pattern 🤔  Just calling `debug` should run an internal check as whether or not it's enabled and output accordingly

Reply:
Ah okay Hannah has said:

> IIRC: debug() is always executed, and then the output is optionally rendered. Meaning that this.query().toQuery() is executed regardless of whether debug is enabled…

So this is to avoid a (presumably) expensive call unless necessary 👍